### PR TITLE
Revamp Notif Logic

### PR DESCRIPTION
**Pull request by vcai122**
Refactors notification logic, allow for multiple tokens per user, restructure to support android (very soon), make views cleaner, etc.

For future PR: fix remaining test cases and management commands.

### CODE REVIEW

File: backend/user/models.py [hunk 1]
```diff
@@ -11,46 +11,52 @@
...
+        abstract = True
 
-    user = models.OneToOneField(User, on_delete=models.CASCADE)
-    kind = models.CharField(max_length=7, choices=KIND_OPTIONS, default=KIND_IOS)
-    token = models.CharField(max_length=255)
+class IOSNotificationToken(NotificationToken):
+    is_dev = models.BooleanField(default=False)
+
+class AndroidNotificationToken(NotificationToken):
+    pass
```
Comment (tuneerroy) [inline backend/user/models.py:23]: would an `is_dev` field not make sense here too? I don't know the Android notif system but would be surprised if they didn't have similar optionality to iOS.

File: backend/user/models.py [hunk 2]
```diff
@@ -11,46 +11,52 @@
...
+#         (SERVICE_UNIVERSITY, "University"),
+#         (SERVICE_LAUNDRY, "Laundry"),
+#     )
 
-    token = models.ForeignKey(NotificationToken, on_delete=models.CASCADE)
-    service = models.CharField(max_length=30, choices=SERVICE_OPTIONS, default=SERVICE_PENN_MOBILE)
-    enabled = models.BooleanField(default=True)
+    
+#     service = models.CharField(max_length=30, choices=SERVICE_OPTIONS, default=SERVICE_PENN_MOBILE)
+#     enabled = models.BooleanField(default=True)
```
Comment (tuneerroy) [inline backend/user/models.py:59]: I assume you'll delete these commented out code once you merge the PR?

File: backend/user/notifications.py [hunk 3]
```diff
@@ -23,57 +23,35 @@
...
-        send_delayed_notifications(tokens, title, body, service, is_dev, is_shadow, delay)
+        send_delayed_notifications(tokens, title, body, category, is_dev, is_shadow, delay)
     else:
-        send_immediate_notifications(tokens, title, body, service, is_dev, is_shadow)
-
-    if not users:  # if to all users, can't be any failed pennkeys
-        return success_users, []
-    failed_users = list(set(users) - set(success_users))
-    return success_users, failed_users
+        send_immediate_notifications(tokens, title, body, category, is_dev, is_shadow)
```
Comment (tuneerroy) [inline backend/user/notifications.py:52]: thoughts on doing:
```
params = (tokens, title, body, category, is_dev, is_shadow)
if delay:
    send_delayed_notifications(*params, delay=delay)
else:
    send_immediate_notifications(*params)
```

File: backend/user/serializers.py [hunk 4]
```diff
@@ -1,40 +1,26 @@
...
+#             token=validated_data["token"], service=validated_data["service"]
+#         ).first()
+#         if setting:
+#             raise serializers.ValidationError(detail={"detail": "Setting already created."})
+#         return super().create(validated_data)
+
+#     def update(self, instance, validated_data):
+#         if instance.service != validated_data["service"]:
+#             raise serializers.ValidationError(detail={"detail": "Cannot change setting service."})
+#         return super().update(instance, validated_data)
```
Comment (tuneerroy) [inline backend/user/serializers.py:23]: just a note so we can delete this code before merge PR?

File: backend/user/views.py [hunk 5]
```diff
@@ -40,64 +40,53 @@ def get_object(self):
         return self.request.user
 
 
-class NotificationTokenView(viewsets.ModelViewSet):
-    """
-    get:
-    Return notification tokens of user.
-    """
-
+class NotificationTokenView(APIView, ABC):
```
Comment (tuneerroy) [inline backend/user/views.py:43]: yoooo ABC lol

File: backend/user/views.py [hunk 6]
```diff
@@ -40,64 +40,53 @@ def get_object(self):
...
 
-        token = NotificationToken.objects.filter(user=user).first()
-        if not token:
-            return Response({"service": pk, "enabled": False})
-        setting, _ = NotificationSetting.objects.get_or_create(token=token, service=pk)
-        return Response(NotificationSettingSerializer(setting).data)
+class NotificationServiceView(APIView):
+    permission_classes = [IsAuthenticated]
+
+    # TODO: this is becoming a common pattern, consider abstracting
```
Comment (tuneerroy) [inline backend/user/views.py:87]: note on the TODO

File: backend/user/views.py [hunk 7]
```diff
@@ -109,27 +98,37 @@ class NotificationAlertView(APIView):
...
         if None in [service, title, body]:
             return Response({"detail": "Missing required parameters."}, status=400)
-        if service not in dict(NotificationSetting.SERVICE_OPTIONS):
+
+        if not (service_obj := NotificationService.objects.filter(name=service).first()):
             return Response({"detail": "Invalid service."}, status=400)
+        
+        users_with_service = service_obj.enabled_users.filter(username__in=usernames)
+
+        tokens = list(IOSNotificationToken.objects.filter(user__in=users_with_service, is_dev=is_dev).values_list(
```
Comment (tuneerroy) [inline backend/user/views.py:121]: just an opinion but probably prettier to extract this into a `get_tokens()` method within the `post` method but I defer to your judgement

**Review by tuneerroy — APPROVED** [review]
Did we want to handle switching users to holding a list of tokens rather than individual tokens in this PR or a later PR? (for ex, user has Android phone + iPad)

File: backend/user/models.py (hunk 1 above)
Comment (vcai122) [reply backend/user/models.py:23]: after talking to android didn't seem like they had

**Comment by vcai122** [comment]
> Did we want to handle switching users to holding a list of tokens rather than individual tokens in this PR or a later PR? (for ex, user has Android phone + iPad)

I think we do right? I changed the user relation on token to be a foreign key.

**Comment by vcai122** [comment]
> Did we want to handle switching users to holding a list of tokens rather than individual tokens in this PR or a later PR? (for ex, user has Android phone + iPad)

I think we do right? I changed the user relation on token to be a foreign key.

**Pull request merged**